### PR TITLE
Decorators 2 Transform [WIP]

### PR DESCRIPTION
**Pull request by peey**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues             | Progress on https://github.com/babel/proposals/issues/11
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added/Pass?        | Yes
| Spec Compliancy?         | Yes
| License                  | MIT
| Doc PR                   | No
| Any Dependency Changes?  | Yes, on decorators2 syntax plugin

Note: I'll use the terms *evaluation* to mean evaluation of `@dec(args)` and *decoration* to mean actual execution of the decorators

Also, note that the decorator proposal introduces the concept of an [*elementDescriptor*](https://tc39.github.io/proposal-decorators/#sec-to-element-descriptor), which is distinct from the [*propertyDescriptor*](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor) we all know about.

To understand the code, you need to know that an *elementDescriptor* is an object with the shape:
```
{
  kind: "property", // I don't know what other values kind can take
  key: <the name of the method>,
  isStatic: true|false,
  descriptor: <the property descriptor of the method>
}
```

I've implemented the methods `decorate`, `decorateElement`, `decorateClass`, `makeElementDescriptor`, and `mergeDuplicateElements` in `helpers.js`. Out of these, `decorateElement` and `decorateClass` correspond to the spec operations [*DecorateElement*](https://tc39.github.io/proposal-decorators/#sec-decorate-element), and [*DecorateClass*](https://tc39.github.io/proposal-decorators/#sec-decorate-class) respectively. `mergeDuplicateElements` corrseponds to the operation used in step 6 of both *DecorateClass* and *DecorateElement*. 

The method `decorate` runs the whole thing. It takes care of the evaluation order (see below) and corrsponds to [Runtime Semantics: ClassDefinitionEvaluation](https://tc39.github.io/proposal-decorators/#sec-runtime-semantics-classdefinitionevaluation) step 21 onwards.

The changes in babel-core are from https://github.com/babel/babel/pull/6058

## Assumptions Made: (Please feel free to correct)
1. evaluation order is as the following: ([relevant spec operation](https://tc39.github.io/proposal-decorators/#sec-runtime-semantics-classdefinitionevaluation))
  a. evaluate all method decorators in order [*step 21.g.* of above spec link] (left to right / outer to inner)
  b. (decoration) execute all method decorators in reverse order *step 24.a.*(right to left / inner to outer)
  c. evaluate all class decorators in order *step 29. (assuming that DecoratorListEvaluation happens just before this, since it's not explicitly mentioned)*
  d. (decoration) execute all class decorators in reverse order *step 29*
2. the signature of decorators is as follows: 
```js
// for member decorator @foo(a,b,c)
function foo(a, b, c) {
   return (descriptor) => {descriptor, extras, finishers};
} 

// for member decorator @foo
function foo(descriptor) {return {descriptor, extras, finishers}}

// for class decorator @foo(a,b,c)
function foo(a,b,c) {
  // elements is an array of  descriptors of all members after running the decorators
  return (ctor, parent, memberDescriptors) => {constructor, elements, finishers}
}

// for class decorator @foo
function (ctor, parent, memberDescriptors) {return {constructor, elements, finishers}}
```
3. The spec uses two forms - a `finisher` attached to the *elementDescriptor* and explicitly returning *finishers* array. I've largely ignored the former because it doesn't seem to make any sense. 
4. The spec is clear that `elementDescriptor` is passed to method decorators and not `propertyDescriptor`, and that's how I had implemented it in the past. However, I changed it to use `propertyDescriptor` because using `elementDescriptor` would imply that the decorator can change the key of the method being decorated, and if it is static or not. I'm not sure if this behaviour is intended or unintended. I need more feedback on this from the spec designers. Changing back to `elementDescriptor` shouldn't be too much trouble.

## TODO / WIP:
 - tests for heritage
 - configurability check and error if a decorator tries to configure a non-configurable descriptor
 - handling undecorated accessors descriptors (getters and setters)
 - more tests for transformation, optimizing for special cases.

### CODE REVIEW

**Comment by peey** [comment]
I accidentally opened this PR with incomplete info. Now that I've edited it to include all details, I'm reopening

**Comment by littledan** [comment]
I'm proposing some changes to the decorator proposal at https://github.com/littledan/proposal-unified-class-features . You can see a summary of the changes in [this presentation](https://docs.google.com/presentation/d/1AOSCuB1TLljBZFMLVlFfXz9IXQAdzC-VzMtyUT9vtnI/edit). The committee and decorators champions seemed to view it positively. I'd be interested to hear what you think.

I don't think this follow-on proposal should affect most of the implementation. I'll review this patch as an implementation of [this decorators proposal](https://github.com/tc39/proposal-decorators).

**Comment by jkrems** [comment]
@littledan Does removing finishers from element decorators mean they can no longer safely attach meta data? Especially if finishers can no replace the class completely? I thought this was done explicitly to allow meta data "annotation"-style decorators that are hard or impossible to implement with stage 0?

**Comment by littledan** [comment]
@jkrems You can attach metadata by putting it on the methods in an additional property, or making a WeakMap keyed off of the method. Would that work for you?

**Comment by jkrems** [comment]
@littledan The nice thing about finalizers was that it allowed to the 2nd thing safely - no matter what other decorators would do.

```js
@injectDataSource
class MyResource {
  @deprecated // not 100% sure about evaluation order
  @GET('/users/{id}') // wants to attach meta data to `getUser`
  async getUser(params) { /* ... */ }
}
```

We're assuming here that:

1. `deprecated` wraps the actual function value that `GET` saw in another function that logs a deprecation warning.
2. `injectDataSource` could be something like a redux connect helper. The point being: It wants to wrap the constructor itself.

In the previous stage 2 iteration, `@GET` could use a finalizer to `myWeakMap.set(finalCtor, metaData)`. If it tries to do the same during its own run, it would get the MyResource class *before* being wrapped. And the meta data would get lost. If it tries to attach the meta data to the function/property value itself, it will be swallowed by `deprecated`.

File: packages/babel-plugin-transform-decorators-2/test/fixtures/examples/logger/exec.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+const calls = [];
+
+function log(message) {
+  return function (descriptor) {
+    let oldFunc = descriptor.value;
+    descriptor.value = function(...args) {
```
Comment (Kovensky) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/examples/logger/exec.js]: Somewhat unfortunately, there is no way of making this function's name derive from the original function name (instead of "value") other than by calling `defineProperty` on it afterwards 🤔

File: packages/babel-plugin-transform-decorators-2/test/fixtures/examples/overrider/exec.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+function overrider(fn) {
+  return function(descriptor) {
+    descriptor.value = fn;
+    return {descriptor, extras: [], finishers: []};
+  }
+}
+
+class Foo {
+  @overrider(() => 3) method() {
```
Comment (Kovensky) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/examples/overrider/exec.js]: What is `this` here? Is it the constructor, the outer `this`, or the newly created instance?

Could be worth having a test for it.

File: packages/babel-plugin-transform-decorators-2/test/fixtures/examples/spare/exec.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// makes an undecorated copy of the method
+function spare(name, isStatic) {
+  return function(descriptor) {
+    const clone = {
+      kind: "property",
+      isStatic: !!isStatic,
+      key: name,
```
Comment (Kovensky) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/examples/spare/exec.js]: What should happen if the name conflicts? `@spare("foo") foo() {}`

File: packages/babel-plugin-transform-decorators-2/test/fixtures/transformations/method-and-class-decorators/expected.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+let Foo = babelHelpers.decorate(class Foo {
+  method() {}
+
+  undecorated() {}
+
+}, [["undecorated"]], [["method", [methDec]]], void 0)([classDec]);
```
Comment (Kovensky) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/transformations/method-and-class-decorators/expected.js]: (Having not looked at the decorate implementation yet) Too many arrays? Some of these look like they could be flattened one level, but I'm only guessing at the meaning of each parameter.

File: packages/babel-plugin-transform-decorators-2/test/fixtures/wip/wip/expected.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  m2() {}
+
+  static [(_key = 3 + 7)]() {}
+
+  andAnUndecoratedMethod() {}
+
+  [calculated + and + undecorated]() {}
+
+}, [["andAnUndecoratedMethod"], [calculated + and + undecorated]], [["m1", [dec]], ["m2", [bar, foo.bar(baz)]], [(_key = 3 + 7), [dec], true]], void 0)([decorator]);
```
Comment (Kovensky) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/wip/wip/expected.js]: `calculated + and + undecorated` should be put in a ref as well (could be an impure expression)

File: packages/babel-helpers/src/helpers.js [hunk 6]
```diff
@@ -270,6 +270,151 @@ helpers.createClass = template(`
...
+          "property",
+          key,
+          isStatic,
+          propertyDescriptor,
+        )
+      );
+    }
+
+    // decorate and store in elementDescriptors
+    for (const [key, decorators, isStatic] of memberDecorators) {
```
Comment (xtuc) [inline packages/babel-helpers/src/helpers.js]: Can we ES2015 here?

And a bit later, `Array.from` will require a polyfill.
It could cause issues like this https://github.com/babel/babel/issues/5876, right?

File: packages/babel-plugin-transform-decorators-2/test/fixtures/wip/wip/expected.js (hunk 5 above)
Comment (peey) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/wip/wip/expected.js]: I'm using babel's built-in `isPure` which is why that wasn't put in a ref

File: packages/babel-helpers/src/helpers.js (hunk 6 above)
Comment (peey) [inline packages/babel-helpers/src/helpers.js]: I wish we could. But if we can't, I'll refactor it to use ES5, it shouldn't be complex. I'll leave this be till the end though, since the code is functionally correct so for the review we could focus on just that

File: packages/babel-plugin-transform-decorators-2/test/fixtures/transformations/method-and-class-decorators/expected.js (hunk 4 above)
Comment (peey) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/transformations/method-and-class-decorators/expected.js]: currently I'm using `const [key, isStatic] = [..]` so that's why the inner items of those are arrays (in case of a non static member, I just omit the second array element). We could certainly simplify it further if needed, e.g. if it's static then it'll look like `["method", true]` otherwise it'll look like just `"method"` and not `["method"]`

File: packages/babel-plugin-transform-es2015-parameters/test/fixtures/parameters/rest-member-expression-optimisation/expected.js [hunk 7]
```diff
@@ -21,6 +21,6 @@ function t() {
 
 function t() {
   for (var i = 0; i < arguments.length; i++) {
-    return arguments.length <= i ? undefined : arguments[i];
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-parameters/test/fixtures/parameters/rest-member-expression-optimisation/expected.js]: this should be fixed

File: packages/babel-plugin-transform-decorators-2/test/fixtures/examples/spare/exec.js (hunk 3 above)
Comment (peey) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/examples/spare/exec.js]: Currently what I've interpreted from the spec is that it should "merge". By that I mean it will just take on the value of the last property added with the key "foo". See [decorateElement](https://tc39.github.io/proposal-decorators/#sec-decorate-element) step 6. Also see `mergeDuplicateElements` helper

File: packages/babel-plugin-transform-decorators-2/test/fixtures/examples/logger/exec.js (hunk 1 above)
Comment (peey) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/examples/logger/exec.js]: Actually, we'll probably be passing the element descriptor to the decorator and not the property descriptor. I just had a doubt on how that'd work when you change the key of the descriptor you've returned and that's why I made the call to implement it like this. Once I get a little clarification on assumption #4 from @littledan or someone, I'll make the desired changes

**Comment by nicolo-ribaudo** [comment]
I just tried this pr locally. `make test-only` works, but I can't succeed in using this plugin 😅 

<details>
<summary>Code</summary>

```js
"use strict";

const babel = require("./packages/babel-core");
const transformDecorators2 = require("./packages/babel-plugin-transform-decorators-2");

const { code } = babel.transform("@foo class Bar {}", {
  plugins: [transformDecorators2],
});

console.log(code);
```
</details>

<details>
<summary>Error</summary>

```
/home/nicolo/Documenti/babel/babel/packages/babel-core/lib/transformation/file/index.js:531   
        err._babel = true;                                                                    
                   ^                                                                          
                                                                                              
TypeError: Cannot create property '_babel' on string 'Helper dependencies must be added to the file first'                                                                                  
    at File.wrap (/home/nicolo/Documenti/babel/babel/packages/babel-core/lib/transformation/file/index.js:531:20)                                                                           
    at Object.transform (/home/nicolo/Documenti/babel/babel/packages/babel-core/lib/transformation/pipeline.js:51:15)                                                                       
    at Object.<anonymous> (/home/nicolo/Documenti/babel/babel/demo.js:6:24)                   
    at Module._compile (module.js:573:30)                                                     
    at Object.Module._extensions..js (module.js:584:10)                                       
    at Module.load (module.js:507:32)                                                         
    at tryModuleLoad (module.js:470:12)                                                       
    at Function.Module._load (module.js:462:3)                                                
    at Function.Module.runMain (module.js:609:10)                                             
    at startup (bootstrap_node.js:158:16)
```
</details>

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+  function addKeysToComputedMembers(path) {
+    const body = path.get("body.body");
+
+    for (const member of body) {
+      if (member.node.computed && member.get("key").isPure()) {
+        const parent = path.findParent(p => p.parentPath.isBlock());
+        const ref = member.scope.generateUidIdentifier("key");
+        parent.insertBefore(
+          t.variableDeclaration("let", [t.variableDeclarator(ref)]),
+        );
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: (Lines 17-21)

Instead of searching the parent block and then adding `let key`, you can use `member.scope.parent.push` which does that for you.

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+/** manual testing code
+**/
+export default function({ types: t }) {
+  // converts [(expression)] to [(let key = (expression))] if expression is impure
+  // so as to avoid recomputation when the key is needed later
+  function addKeysToComputedMembers(path) {
+    const body = path.get("body.body");
+
+    for (const member of body) {
+      if (member.node.computed && member.get("key").isPure()) {
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: I think this should be if `!isPure()`?

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        const parent = path.findParent(p => p.parentPath.isBlock());
+        const ref = member.scope.generateUidIdentifier("key");
+        parent.insertBefore(
+          t.variableDeclaration("let", [t.variableDeclarator(ref)]),
+        );
+        const replacement = t.sequenceExpression([
+          t.assignmentExpression("=", ref, member.node.key),
+        ]);
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: Why is a SequenceExpression needed if there is only one expression?

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    return t.arrayExpression(result);
+  }
+
+  return {
+    inherits: syntaxDecorators2,
+    visitor: {
+      // export default is a special case since it expects and expression rather than a declaration
+      ExportDefaultDeclaration(path) {
+        const classPath = path.get("declaration");
+        if (!classPath.isClassDeclaration()) return;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: I think we can also return if the class doesn't have any decorator?

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            t.assignmentExpression("=", ref, t.toExpression(classPath.node)),
+          );
+        } else {
+          classPath.replaceWith(t.toExpression(classPath));
+        }
+      },
+
+      // replace declaration with a let declaration
+      ClassDeclaration(path) {
+        const { node } = path;
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: > I think we can also return if the class doesn't have any decorator?

Also here

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            t.variableDeclarator(ref, t.toExpression(node)),
+          ]),
+        );
+      },
+
+      ClassExpression(path, file) {
+        file.addHelper("makeElementDescriptor");
+        file.addHelper("mergeDuplicateElements");
+        file.addHelper("decorateClass", ["mergeDuplicateElements"]);
+        file.addHelper("decorateElement", ["mergeDuplicateElements"]);
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: It would be better to include these helpers only if the class has some decorators.

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        const decorateIdentifier = file.addHelper("decorate", [
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: `decorate` should have `makeElementDescriptor` as a dependency.

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+export default function({ types: t }) {
+  // converts [(expression)] to [(let key = (expression))] if expression is impure
+  // so as to avoid recomputation when the key is needed later
+  function addKeysToComputedMembers(path) {
+    const body = path.get("body.body");
+
+    for (const member of body) {
+      if (member.node.computed && member.get("key").isPure()) {
+        const parent = path.findParent(p => p.parentPath.isBlock());
+        const ref = member.scope.generateUidIdentifier("key");
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: IIRC there is a method which generates an uid using a node (e.g. `computed.property + 1` -> `computed$propertyPlus1`), so we have more descriptive names and avoid `_key26739`

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+      ClassExpression(path, file) {
+        file.addHelper("makeElementDescriptor");
+        file.addHelper("mergeDuplicateElements");
+        file.addHelper("decorateClass", ["mergeDuplicateElements"]);
+        file.addHelper("decorateElement", ["mergeDuplicateElements"]);
+        const decorateIdentifier = file.addHelper("decorate", [
+          "decorateClass",
+          "decorateElement",
+        ]);
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: `decorate` should have `makeElementDescriptor` as a dependency.

Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: Also probably dependencies should be declared in the helper definition.

Something like
```js
helpers.myHelper = template("...");
helpers.myHelper.dependencies = ["foo", "bar"];
```

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+    if (node.computed) {
+      // if it is computed, it has been processed by addKeysToComputedMembers
+      if (t.isAssignmentExpression(node.key)) {
+        return node.key.left;
+      } else {
+        // it's pure
+        return node.key;
+      }
+    } else {
+      return t.stringLiteral(node.key.name);
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: The key could also be a number; in this case `key.name` is `undefined`.

```js
class Foo {
  @dec 1() {}
}
```

File: packages/babel-plugin-transform-es2015-parameters/test/fixtures/parameters/rest-member-expression-optimisation/expected.js (hunk 7 above)
Comment (peey) [inline packages/babel-plugin-transform-es2015-parameters/test/fixtures/parameters/rest-member-expression-optimisation/expected.js]: I'm not sure if this is related to this PR. I'll investigate more

File: packages/babel-plugin-transform-decorators-2/test/fixtures/wip/wip/expected.js (hunk 5 above)
Comment (peey) [inline packages/babel-plugin-transform-decorators-2/test/fixtures/wip/wip/expected.js]: You're right, this should have been put in a ref. [Nicolo pointed out the error in code](https://github.com/babel/babel/pull/6107#discussion_r133581559) and I've fixed it, will push soon

File: packages/babel-core/src/transformation/file/index.js [hunk 18]
```diff
@@ -241,7 +240,18 @@ export default class File extends Store {
...
     }
 
-    const ref = getHelper(name);
+    const opts = {};
+    if (deps) {
+      for (const dep in deps) {
+        if (!this.usedHelpers[name]) {
+          throw "Helper dependencies must be added to the file first";
+        } else {
+          opts["babelHelpers." + dep] = this.usedHelpers[name];
```
Comment (peey) [inline packages/babel-core/src/transformation/file/index.js]: Hey. I completely missed this, I guess since it was working on babel tests I assumed it would be generating the correct code. Apparently not. Anyways, I'll track the issues with this feature separately in https://github.com/babel/babel/pull/6058. Also I don't intend to implement too complex features for this - just enough to make it work, because #5706 already is aimed at providing a superset of the functionality

**Comment by peey** [comment]
Current Status of this PR: Couple of updates about what the commits in the past few days have been about

1. Improved the transform by the feedback I got
2. Decorators now get the elementDescriptor and not the propertyDescriptor
3. A question about getters and setters: Step 22 in [decorate class](https://tc39.github.io/proposal-decorators/#sec-decorate-class) seems to be saying that if both getter and setter with the same name have a decorator, we should throw. A couple of questions about this
  a. Is it supposed to throw when decorators for either getter or setter are present or only when both are? If the latter, then what's the rationale behind this?
  b. If The current parser behaviour is that babylon throws a syntax error at compile time if there are decorators on a  getter or setter. Is it supposed to throw at runtime or result in a syntax error? If it's supposed to throw a runtime error I'll send a PR to change that behaviour in babylon.
4. If a deeper decorator sets configurable to false for a property, then the outer decorators aren't allowed to modify the descriptor.
 
Still To Do:
 - tests for heritage. I actually need help with this. What kind of heritage changes are decorators allowed to make? We'll need to add tests for each. What are some good examples of using decorators with a class heritage? (we should test this implementation on these examples)

File: packages/babel-helpers/src/helpers.js [hunk 19]
```diff
@@ -417,6 +562,30 @@ helpers.interopRequireWildcard = template(`
   })
 `);
 
+helpers.makeElementDescriptor = template(`
+  (function (kind, key, isStatic, descriptor, finisher) {
+    return { kind, key, isStatic, descriptor, finisher };
```
Comment (littledan) [inline packages/babel-helpers/src/helpers.js]: Minor, but the finisher property is only created if it is not undefined.

File: packages/babel-helpers/src/helpers.js [hunk 20]
```diff
@@ -417,6 +562,30 @@ helpers.interopRequireWildcard = template(`
...
+  });
+`);
+
+//TODO
```
Comment (littledan) [inline packages/babel-helpers/src/helpers.js]: Would be nice to include in the comment here that the details of merging are still up for discussion in TC39.

Comment (littledan) [reply packages/babel-helpers/src/helpers.js]: The merging here seems like just overwriting and choosing the last one. I think merging should also include merging getter/setter pairs.

File: packages/babel-plugin-transform-decorators-2/src/index.js [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+
+        const undecorated = undecoratedMethods(path);
+        const methodDecorators = takeMethodDecorators(path);
+        const classDecorators = takeClassDecorators(path);
+
+        if (
+          methodDecorators.elements.length == 0 &&
+          classDecorators.elements.length == 0
+        ) {
+          return;
```
Comment (littledan) [inline packages/babel-plugin-transform-decorators-2/src/index.js]: When will this be true when hasDecorators is already true?

File: packages/babel-helpers/src/helpers.js [hunk 22]
```diff
@@ -270,6 +270,151 @@ helpers.createClass = template(`
...
+          propertyDescriptor,
+        )
+      );
+    }
+
+    // decorate and store in elementDescriptors
+    for (const [key, decorators, isStatic] of memberDecorators) {
+      const target = isStatic ? constructor : prototype;
+      const propertyDescriptor =
+        elementDescriptors.has([key, isStatic]) && elementDescriptors.get([key, isStatic]).descriptor
```
Comment (littledan) [inline packages/babel-helpers/src/helpers.js]: This has call will always return false, since ES6 maps are by identity. Since [1, 2] !== [1, 2], calling Map.prototype.has on a fresh array will never do what you're hoping for.

Instead, you could maintain two Sets of keys, or two objects with keys and values, for the static and non-static halves.

File: packages/babel-helpers/src/helpers.js [hunk 23]
```diff
@@ -270,6 +270,170 @@ helpers.createClass = template(`
...
+      const result = babelHelpers.decorateClass(
+        constructor,
+        classDecorators,
+        heritage,
+        Array.from(elementDescriptors.values())
+      );
+
+      finishers = finishers.concat(result.finishers);
+      //TODO: heritage hacks so result.constructor has the correct prototype and instanceof results
+      //TODO: step 38 and 39, what do they mean "initialize"?
```
Comment (littledan) [inline packages/babel-helpers/src/helpers.js]: "Initialize" was about having methods and the constructor start in some state where they would be unusable, and make them usable later. In [my follow-on decorators proposal](https://github.com/littledan/proposal-unified-class-features), I'm suggesting to get rid of this initialized/uninitialized state. I think it's OK to go without it to start, even if it's possible that TC39 will ask to bring it back.

File: packages/babel-helpers/src/helpers.js [hunk 24]
```diff
@@ -270,6 +270,170 @@ helpers.createClass = template(`
...
+    return function(classDecorators) {
+      const result = babelHelpers.decorateClass(
+        constructor,
+        classDecorators,
+        heritage,
+        Array.from(elementDescriptors.values())
+      );
+
+      finishers = finishers.concat(result.finishers);
+      //TODO: heritage hacks so result.constructor has the correct prototype and instanceof results
```
Comment (littledan) [inline packages/babel-helpers/src/helpers.js]: Heritage is just given as an argument to class decorators; it's the responsibility of the decorators to do the appropriate thing with that, I believe.

File: packages/babel-helpers/src/helpers.js (hunk 20 above)
Comment (peey) [reply packages/babel-helpers/src/helpers.js]: @littledan, I have a question. Does merging also mean that if by a previous decorator you have some attribute, then if it isn't provided by a later decorator with an extra for the same key, then we just use the attribute from the previous extra object? For example, if enumerability is set to `false` for an extra with key "foo" by a previous decorator, then a later decorator needn't provide a enumerability value for an extra with key "foo", and we'll just use the previous enumerability value?

Current behaviour in this PR expects that all extras are well-formed descriptors, which means that we simply overwrite and choose the last one, as you noticed.

**Comment by peey** [comment]
Apologies for responding late to the review. Uni workload has started again. 

Updates: Now we're merging getters and setters & other things pointed out by littledan in their review

Todo:
- Refactor merging so we can merge when adding extras instead of overriding
- Look into if other props things like enumerable, configurable, writable & value (in case of data descriptor) are also meant to be merged. @littledan thoughts on this?
- Fix the failing test for `x instanceof Foo` if `Foo` has been processed by the transform

**Comment by peey** [comment]
Update: @nicolo-ribaudo asked me if they could take this up, and I told them they could. 

The outstanding work on this PR would have been the todo in [my previous comment](https://github.com/babel/babel/pull/6107#issuecomment-329958982) and then any changes between the spec at the time this PR was made and the spec now (I'm not sure about the exact status of this, seems like http://tc39.github.io/proposal-unified-class-features  that @littledan's [comment](https://github.com/babel/babel/pull/6107#pullrequestreview-64871425) specified is dead and  http://tc39.github.io/proposal-decorators has been updated)

Since Babel's codebase has changed much (and the proposal may have too), I guess Nicolo's taking the route of reworking this feature and salvaging code from this PR as needed in https://github.com/babel/babel/pull/7542 . Kudos and good luck!

Here's what might be useful from this PR:
 - the file [babel-plugin-transform-decorators-2/src/index.js](https://github.com/peey/babel/blob/76f5d3cca8d00e23bed36c418246fdec1ef0ba04/packages/babel-plugin-transform-decorators-2/src/index.js) houses the transformation logic
 - changes to [helpers.js](https://github.com/babel/babel/pull/6107/files#diff-ba75bd4d116cc90ad8c8002e3010e821) have the functionally equivalent javascript code to emulate decorators
 - the [test folder](https://github.com/peey/babel/tree/76f5d3cca8d00e23bed36c418246fdec1ef0ba04/packages/babel-plugin-transform-decorators-2/test) has numerous corner cases, and decorator usage examples

**Comment by nicolo-ribaudo** [comment]
Closing in favor #7976
